### PR TITLE
feat: add a CI pipeline for publishing docker images

### LINKED ISSUE #292: Add CI support for publishing docker container images
### Problem

When new Hedera Node tags are delivered, the container images must be manually built and released.

### Solution

- Add the docker container definitions to the repository.
- Add a CI pipeline to the repository capable of building and publishing the images.
- Add self-hosted CI runner support to this repository.
- Add a trigger in the `hashgraph/hedera-services` repository to initiate the workflow when a tag is created.

### Alternatives

None

### PR DESCRIPTION
**Pull request by nathanklick**
## Description

Adds a CI pipeline to publish and deploy docker images for official tagged builds of the Hedera Node software.

### Related Issues

- Closes #292

### CODE REVIEW

File: .github/workflows/zxf-publish-docker-images.yaml [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+    steps:
+      - name: Render Job Summary
+        run: |
+          printf "### Published Docker Images\n" >> "${GITHUB_STEP_SUMMARY}"
+          printf "| Image Name | Version | URL | Supported Architectures |\n" >> "${GITHUB_STEP_SUMMARY}"
+          printf "| ---------- | ------- | --- | ----------------------- |\n" >> "${GITHUB_STEP_SUMMARY}"
+          printf "| %s | %s | %s | %s |\n" \
+            "gcr.io/hedera-regsitry/network-node-base" \
+            "${{ needs.prepare-release.outputs.version }}" \
+            "[GCP Console](https://console.cloud.google.com/gcr/images/hedera-registry/global/network-node-base?project=hedera-registry)" \
```
Comment (isavov) [inline .github/workflows/zxf-publish-docker-images.yaml:246]: `?project=hedera-registry` this part of the url forces users to be logged in to google to access it
`gcr.io/hedera-registry/network-node-base` is a shorter url that is accessible for everyone

File: .github/workflows/zxf-publish-docker-images.yaml [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+      - deploy-network-node-image
+      - deploy-haveged-image
+    steps:
+      - name: Render Job Summary
+        run: |
+          printf "### Published Docker Images\n" >> "${GITHUB_STEP_SUMMARY}"
+          printf "| Image Name | Version | URL | Supported Architectures |\n" >> "${GITHUB_STEP_SUMMARY}"
+          printf "| ---------- | ------- | --- | ----------------------- |\n" >> "${GITHUB_STEP_SUMMARY}"
+          printf "| %s | %s | %s | %s |\n" \
+            "gcr.io/hedera-regsitry/network-node-base" \
```
Comment (isavov) [inline .github/workflows/zxf-publish-docker-images.yaml:244]: typo - should be `registry`, image is unreachable

File: .github/workflows/zxf-publish-docker-images.yaml [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+            "${{ needs.prepare-release.outputs.version }}" \
+            "[GCP Console](https://console.cloud.google.com/gcr/images/hedera-registry/global/network-node-base?project=hedera-registry)" \
+            "linux/amd64, linux/arm64" >> "${GITHUB_STEP_SUMMARY}"
+          
+          printf "| %s | %s | %s | %s |\n" \
+            "gcr.io/hedera-regsitry/network-node-haveged" \
```
Comment (isavov) [inline .github/workflows/zxf-publish-docker-images.yaml:250]: typo is propagated

File: .github/workflows/zxf-publish-docker-images.yaml [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+            "${{ needs.prepare-release.outputs.version }}" \
+            "[GCP Console](https://console.cloud.google.com/gcr/images/hedera-registry/global/network-node-haveged?project=hedera-registry)" \
+            "linux/amd64, linux/arm64" >> "${GITHUB_STEP_SUMMARY}"
+          
+          printf "| %s | %s | %s | %s |\n" \
+            "gcr.io/hedera-regsitry/main-network-node" \
```
Comment (isavov) [inline .github/workflows/zxf-publish-docker-images.yaml:256]: typo is propagated

**Comment by isavov** [comment]
I have a general question about this CI. Isn't this workflow better suited for the `hedera-services` repo? It seems to me that the images created are not specific to local node, are they needed only by it?

**Review by Nana-EC — APPROVED** [review]
Awesome. Beyond Illiya's comments LGTM.
Might be good to clarify with additional README instructions. If this works let's use it and iterate on it.
Thanks

**Review by jeromy-cannon — APPROVED** [review]

File: .github/workflows/zxf-publish-docker-images.yaml (hunk 1 above)
Comment (nathanklick) [reply .github/workflows/zxf-publish-docker-images.yaml:246]: This is the URL to the GCP Console and does require authentication.

**Comment by nathanklick** [comment]
> I have a general question about this CI. Isn't this workflow better suited for the `hedera-services` repo? It seems to me that the images created are not specific to local node, are they needed only by it?

Actually these image definitions are specific to local node and have been for a long time. Actual network deployments use a different image definition.

**Comment by nathanklick** [comment]
> Awesome. Beyond Illiya's comments LGTM. Might be good to clarify with additional README instructions. If this works let's use it and iterate on it. Thanks

@Nana-EC I am not sure what instruction would need to be in a public readme. This is part of an automated process and should not be used by any user directly.

File: .github/workflows/zxf-publish-docker-images.yaml (hunk 2 above)
Comment (nathanklick) [reply .github/workflows/zxf-publish-docker-images.yaml:244]: Fixed

File: .github/workflows/zxf-publish-docker-images.yaml (hunk 4 above)
Comment (nathanklick) [reply .github/workflows/zxf-publish-docker-images.yaml:256]: Fixed

File: .github/workflows/zxf-publish-docker-images.yaml (hunk 1 above)
Comment (nathanklick) [reply .github/workflows/zxf-publish-docker-images.yaml:246]: Working as intended, no change required.

Comment (nathanklick) [reply .github/workflows/zxf-publish-docker-images.yaml]: Removing the `?project=hedera-registry` portion of the URL does remove the need to authenticate to GCP. However, the original intent was to provide a link for our engineers to quickly access the portal to manage images. This pipeline is a derivative of the pipeline currently in use by the `hashgraph/hedera-mirror-node-explorer` repository and is not intended for use by users in either case.

Comment (nathanklick) [reply .github/workflows/zxf-publish-docker-images.yaml]: I went ahead and removed the query string.

**Comment by Nana-EC** [comment]
> > Awesome. Beyond Illiya's comments LGTM. Might be good to clarify with additional README instructions. If this works let's use it and iterate on it. Thanks
> 
> @Nana-EC I am not sure what instructions would need to be in a public readme. This is part of an automated process and should not be used by any user directly.

Okay great. I wasn't clear if there were any manual steps that could be taken. 
Automated is great. Thanks

**Pull request merged**